### ISSUE TITLE
Attach data to hg_bulk_t and hg_addr_t

### ISSUE THREAD

**Issue by mdorier** [mercury-hpc/mercury#627]
**Is your feature request related to a problem? Please describe.**

In Margo we use `HG_Set/Get_data` to attach data to `hg_handle_t` handles (typically we attach the `margo_instance_id` that was used to create the handle, along some other information). A similar mechanism for `hg_bulk_t` and `hg_addr_t` would be really useful.

**Describe the solution you'd like**

```
hg_return_t HG_Bulk_set_data(hg_bulk_t handle, void *data, void (*free_callback)(void *));
void * HG_Bulk_get_data(hg_bulk_t handle);

hg_return_t HG_Addr_set_data(hg_addr_t handle, void *data, void (*free_callback)(void *));
void * HG_Addr_get_data(hg_addr_t handle);
```

**Comment by soumagne** [comment]
Yes I think I agree and that would prevent you for example from re-implementing another hash table or something similar to retrieve whatever you need to associate to the addr / bulk.
